bindings/rust/s2n-tls/src/raw/config.rs

```diff
@@ -0,0 +1,180 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::raw::error::Error;
+use alloc::rc::Rc;
+use core::convert::TryInto;
+use s2n_tls_sys::*;
+use std::ffi::CString;
+
+struct Owned(*mut s2n_config);
+
+impl Default for Owned {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Owned {
+    fn new() -> Self {
+        crate::raw::init::init();
+        let config = call!(s2n_config_new()).unwrap();
+        Self(config)
+    }
+
+    pub(crate) fn as_mut_ptr(&mut self) -> *mut s2n_config {
+        self.0
+    }
+}
+
+impl Drop for Owned {
+    fn drop(&mut self) {
+        let _ = call!(s2n_config_free(self.0));
+    }
+}
+
+#[derive(Clone, Default)]
+pub struct Config(Rc<Owned>);
+
+impl Config {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn builder() -> Builder {
+        Builder::default()
+    }
+
+    pub(crate) fn as_mut_ptr(&mut self) -> *mut s2n_config {
+        (self.0).0
+    }
+}
+
+#[derive(Default)]
+pub struct Builder(Owned);
+
+impl Builder {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    pub fn set_alert_behavior(
+        &mut self,
+        value: s2n_alert_behavior::Type,
+    ) -> Result<&mut Self, Error> {
+        call!(s2n_config_set_alert_behavior(self.as_mut_ptr(), value))?;
+        Ok(self)
+    }
+
+    pub fn set_cipher_preference(&mut self, name: &str) -> Result<&mut Self, Error> {
+        let name = CString::new(name).map_err(|_| Error::InvalidInput)?;
+        call!(s2n_config_set_cipher_preferences(
+            self.as_mut_ptr(),
+            name.as_ptr() as *const _
+        ))?;
+        Ok(self)
+    }
+
+    /// sets the application protocol preferences on an s2n_config object.
+    ///
+    /// protocols is a list in order of preference, with most preferred protocol first,
+    /// and of length protocol_count. When acting as an S2N_CLIENT the protocol list is
+    /// included in the Client Hello message as the ALPN extension. As an S2N_SERVER, the
+    /// list is used to negotiate a mutual application protocol with the client. After
+    /// the negotiation for the connection has completed, the agreed upon protocol can
+    /// be retrieved with s2n_get_application_protocol
+    pub fn set_alpn_preference<P: IntoIterator<Item = I>, I: AsRef<[u8]>>(
+        &mut self,
+        protocols: P,
+    ) -> Result<(), Error> {
```

Comment:
```suggestion
    ) -> Result<&mut Self, Error> {
```

Reply:
not sure what the call pattern is for this but other fn in this file return Self

Reply:
it's so you can chain setters

```rust
Config::new()
    .set_alpn_preferences(&["h2", "h3"])
    .set_security_preference("default");
```

It's good to be consistent so good catch.

Reply:
the `connection.rs` fns were not returning `Self` but again depending on the call pattern maybe we return Self.